src/main/scala/ir/Program.scala

```diff
@@ -111,25 +170,105 @@ class Program(var procedures: ArrayBuffer[Procedure], var mainProcedure: Procedu
 
 }
 
-class Procedure(
-    var name: String,
-    var address: Option[Int],
-    var blocks: ArrayBuffer[Block],
-    var in: ArrayBuffer[Parameter],
-    var out: ArrayBuffer[Parameter]
-) {
-  def calls: Set[Procedure] = blocks.flatMap(_.calls).toSet
+class Procedure private (
+                  var name: String,
+                  var address: Option[Int],
+                  var entryBlock: Option[Block],
+                  private val _blocks: mutable.HashSet[Block],
```

Comment:
What's the point of making this a set? This seems like it will only cause problems for us, since we want to output blocks in a consistent order.

Reply:
My understanding is the block order does not have any meaning beyond the first block. The set removes the notion of an index at all and gives removal using the object identity instead, while also being faster. But I guess that only matters under heavy modification.

Reply:
LinkedHashSet would be another good option yeah

Reply:
Although that still requires storing the entry block separately which still makes creating the iterator awkward, so I'd really just lean towards ListBuffer. I guess most cases where we iterate over all blocks don't care whether the entry block is first, except the final output, so taking that into account would allow for other possibilities.